pkg/uki/upgrade.go

```diff
@@ -34,19 +40,62 @@ func (i *UpgradeAction) Run() (err error) {
 		return err
 	}
 	cleanup.Push(umount)
-	// TODO: Check size of EFI partition to see if we can upgrade
-	// TODO: Check size of source to see if we can upgrade
 	// TODO: Check number of existing UKI files
-	// TODO: Load them, order them via semver
-	// TODO: Remove the latest one if its over the max number of entries
+	efiFiles, err := i.getEfiFiles()
+	if err != nil {
+		return err
+	}
+	i.cfg.Logger.Infof("Found %d UKI files", len(efiFiles))
+	i.cfg.UkiMaxEntries = 1
+	if len(efiFiles) == i.cfg.UkiMaxEntries {
+		i.cfg.Logger.Infof("Found %d UKI files, which is over max entries allowed(%d) removing the oldest one", len(efiFiles), i.cfg.UkiMaxEntries)
+		versionList := semver.Collection{}
+		for _, f := range efiFiles {
+			versionList = append(versionList, semver.MustParse(f))
+		}
+		// Sort it so the oldest one is first
+		sort.Sort(versionList)
+		i.cfg.Logger.Debugf("All versions found: %s", litter.Sdump(versionList))
+		// Remove the oldest one
+		i.cfg.Logger.Infof("Removing: %s", filepath.Join(i.spec.EfiPartition.MountPoint, "EFI", "kairos", versionList[0].Original()))
+		err = i.cfg.Fs.Remove(filepath.Join(i.spec.EfiPartition.MountPoint, "EFI", "kairos", versionList[0].Original()))
+		if err != nil {
+			return err
+		}
+		// Remove the conf file as well
+		i.cfg.Logger.Infof("Removing: %s", filepath.Join(i.spec.EfiPartition.MountPoint, "loader", "entries", versionList[0].String()+".conf"))
+		// Don't care about errors here, systemd-boot will ignore any configs if it cant find the efi file mentioned in it
+		e := i.cfg.Fs.Remove(filepath.Join(i.spec.EfiPartition.MountPoint, "loader", "entries", versionList[0].String()+".conf"))
+		if e != nil {
+			i.cfg.Logger.Warnf("Failed to remove conf file: %s", e)
+		}
+	} else {
+		i.cfg.Logger.Infof("Found %d UKI files, which is under max entries allowed(%d) not removing any", len(efiFiles), i.cfg.UkiMaxEntries)
+	}
+
 	// Dump artifact to efi dir
 	_, err = e.DumpSource(constants.UkiEfiDir, i.spec.Active.Source)
 	if err != nil {
 		return err
 	}
-	
+
 	_ = elementalUtils.RunStage(i.cfg, "kairos-uki-upgrade.after")
 	_ = events.RunHookScript("/usr/bin/kairos-agent.uki.upgrade.after.hook") //nolint:errcheck
 
 	return hook.Run(*i.cfg, i.spec, hook.AfterUkiUpgrade...)
 }
+
+func (i *UpgradeAction) getEfiFiles() ([]string, error) {
+	var efiFiles []string
+	files, err := os.ReadDir(filepath.Join(i.spec.EfiPartition.MountPoint, "EFI", "kairos"))
```

Comment:
I wonder if the function getEfiFile should belong to the UpgradeAction or to the MountPoint, I'd be more inclined to the latter seeing this here (it would then need to be exported, but still feels more likely to me that I would want the efi files from a mount point than from an upgrade action so to speak.

Reply:
yes, I was struggling on where to put this here, also thougth about making it part of the fsutils somehow but didnt want to start playing with the imports here and there xD

Also notice that the EfiPartition is a very generic struct(v1.Partition) that works for everything else, so having it part of that felt a bit wrong as it was not generic enough as to be valid for anything else in the agent. Is very very specific to uki and to the upgrade part.